Code_Exercises/CMakeLists.txt

```diff
@@ -8,32 +8,102 @@
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_LANGUAGE_VERSION=2020)
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_ACADEMY_USE_COMPUTECPP)
   target_include_directories("${prefix}_${source}" PRIVATE
-    ${PROJECT_SOURCE_DIR}/Utilities/include ${PROJECT_SOURCE_DIR}/External/stb)
-  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads)
+    ${PROJECT_SOURCE_DIR}/Utilities/include 
+    ${PROJECT_SOURCE_DIR}/External/stb)
+  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads) 
   target_link_libraries("${prefix}_${source}" PUBLIC Catch2::Catch2)
-  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17
-    CXX_STANDARD_REQUIRED ON)
-
-  add_sycl_to_target(
-    TARGET "${prefix}_${source}"
-    SOURCES "${source}.cpp"
-  )
-    
+  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)  
   add_test(${prefix}_${source} ${prefix}_${source})
 endfunction()
 
-function(add_sycl_executable_dpcpp prefix source)
-  # Not implemented
+# CMake function to build a hipSYCL configuration and create executable binaries
+# from hipSYCL libraries. 
+function( add_sycl_executable_hipSYCL prefix source )
+  if(NOT DEFINED ${CMAKE_BUILD_TYPE})
+	  set(CMAKE_BUILD_TYPE "DEBUG")
+  endif()
+
+  add_executable("${prefix}_${source}" "${source}.cpp")
+  target_compile_definitions("${prefix}_${source}" PUBLIC -DSYCL_LANGUAGE_VERSION=2020)
+  target_include_directories("${prefix}_${source}" PRIVATE
+    ${PROJECT_SOURCE_DIR}/Utilities/include 
+    ${PROJECT_SOURCE_DIR}/External/stb)
+  target_link_libraries("${prefix}_${source}" PRIVATE Threads::Threads) 
+  target_link_libraries("${prefix}_${source}" PUBLIC Catch2::Catch2)
+  set_target_properties("${prefix}_${source}" PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)  
+endfunction()
+
+# CMake function to build a DPC++ configuration and create executable binaries
+# from oneAPI libraries. 
+function( add_sycl_executable_dpcpp prefix source )
+  if(NOT DEFINED ${CMAKE_BUILD_TYPE})
+    set(CMAKE_BUILD_TYPE "DEBUG")
+  endif()
+
+  set(CMAKE_VERBOSE_MAKEFILE ON)
+  set(TARGET_NAME "${prefix}_${source}")
+  set(CMAKE_CXX_STANDARD 17)
+  set(CMAKE_CXX_COMPILER "icpx")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl")
+  if(CMAKE_BUILD_TYPE STREQUAL "DEBUG")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0")
+  else()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2")
+  endif()
+  
+  add_executable(${TARGET_NAME} "${source}.cpp")
+  # This function is from IntelSYCL cmake module
+  add_sycl_to_target(TARGET ${TARGET_NAME} SOURCES "${source}.cpp") 
+  target_include_directories(${TARGET_NAME} PUBLIC "${CMAKE_SOURCE_DIR}/External/Catch2/single_include")
+  target_link_libraries(${TARGET_NAME} PUBLIC Catch2::Catch2)
 endfunction()
 
+function( add_sycl_executable exercise_number prefix source )
```

Comment:
If we add targets for all the exercises, and unify the build functions, then we can make this function disappear.

Reply:
I do not understand 'add targets for all the exercises'. What do you mean by 'build functions'?

Reply:
I'm referring to [this comment](https://github.com/codeplaysoftware/syclacademy/pull/273/files#r1358418977). It's quite unusual to allow the user to control exactly which targets are created to this degree, I would prefer instead that all targets are created every time and the user can then build whichever target they want with `ninja exercise_02` or similar.

By build functions, I mean the `add_sycl_executable_(dpcpp|hipsycl|computecpp)` functions. If we unify them into one function, and don't let the user choose which exercises are configured, we can remove all this logic (which is simpler).

Reply:
The ability to build individual exercises is what was asked for when making the improvements. You are proposing to reverse all the work that has been done. You did not state how it should be done. Ninja was not mentioned. Any changes ATM are being performed across two branches because of dependencies. It this new approach is needed then it can be done as a new task after the syncing of the branches and in a stable state.